### DISCUSSION TITLE
Mujoco impratio

### DISCUSSION THREAD

**Discussion by Aristomo** [google-deepmind/mujoco#656] [Asking for Help]
In the documents of Mujoco, they say "impratio: This attribute determines the ratio of frictional-to-normal constraint impedance for elliptic friction cones. The setting of solimp determines a single impedance value for all contact dimensions, which is then modulated by this attribute. Settings larger than 1 cause friction forces to be “harder” than normal forces, having the general effect of preventing slip, without increasing the actual friction coefficient. For pyramidal friction cones the situation is more complex because the pyramidal approximation mixes normal and frictional dimensions within each basis vector; but the overall effect of this attribute is qualitatively similar."

I still cannot understand what its meaning. What is the meaning of harder? If I can adjust this value to prevent slipping, why this will not affect the friction coefficient? How can I set the friction coefficient between two bodies? And what are the differences between the 'impartio' and the parameter controling the friction coefficient?

Thanks for your patient response! I am waiting for you. Your response will be very helpful. Thanks again.

**Comment by Balint-H** [comment]
User here, so take my answer with salt. My understanding is that this relates to the concept of switching a contact from "static" mode to "sliding" mode (when the contact force exits the friction cone/pyramid). So with a high impratio but low friction coefficient you can prevent slips, but once a slip happens you will slide along easily. [This page](https://scaron.info/robot-locomotion/friction-cones.html) seems to provide a good description of the concept.

**Reply by Aristomo** [reply]
Thank you so much to response me. So, impratio is like 'static friction coefficient'?

**Reply by Balint-H** [reply]
I think impratio scales the static friction coefficient in comparison to the sliding one. So I believe that at impratio==1, the static and sliding friction coefficients are the same. I'm not sure how exactly the solimp parameters influence this. I suppose since they are involved with solving contacts and determining contact forces it might refer to that? I'd appreciate if someone could correct or confirm this!

**Reply by Aristomo** [reply]
Thanks. I understand your meaning, which is very helpful. I will do some experiments to verify it.

**Comment by yuvaltassa** [comment]
It is a little bit difficult to explain the meaning of `impratio`. My top level recommendation is to gain understanding by playing with a specific model. Here is my "hand wavy" explanation:

1. `solimp`, the "impedance" of a constraint, roughly means "the solver's ability to prevent constraint violation, as prescribed by `solref`". Note the difference here, `solref` roughly means what the constraint "wants to do", as opposed to `solimp` which is what the solver "is able to do". So decreasing `solimp` down from 1 towards 0, decreases the solver's **ability**, but by doing so it also makes the optimisation easier to solve, improves the quality of the solution, allows bigger timesteps etc (not to mention improving realism since in the real world nothing is infinitely stiff).

2. For _elliptic_ friction cones, the normal force and friction force are solved for simultaneously, which means that softening the constraint by reducing `solimp` also has the effect of reducing friction, leading to slip. However this is often _the opposite_ of what you want. For example when modeling rubbery feet, low `solimp` is the the correct way of capturing the softness of the rubber (where penetration of the foot is capturing the deformation, but without the expensive deformation DoFs), yet rubber is often sticky and the softness should not lead to slip. In this case one increases `impratio` which roughly means "the ratio of the solver's ability to prevent sliding to the solver's ability to prevent penetration".

I recommend playing with e.g. the [ANYmal](https://mujoco.readthedocs.io/en/latest/models.html#bipeds-quadrupeds) models which have sticky-rubbery feet, and the above example is relevant (note the [high impratio](https://github.com/deepmind/mujoco_menagerie/blob/d98292efc73511aa7a4ca958eaaf226403d56cb7/anybotics_anymal_b/anymal_b.xml#L4)). Remember that `simulate` has a "contact override" feature which lets you set the contact parameters globally for all contacts, which should help you gain an intuition.

Hope this helps.

**Reply by Balint-H** [reply]
Thanks for the breakdown! I'll try to play around with it to get a sense of it.

Until then, could you please give a bit of extra detail on static friction? Is there an explicit static friction (that is higher than sliding/kinetic friction) simulated? If yes, then does increasing impratio increase static friction only or also sliding friction? Lastly does a compliant solimp lower the effective sliding friction, or only the static friction?
